seo/builder.go

```diff
@@ -0,0 +1,373 @@
+		}
+		b.registeredSEO[modelType] = seo
+	}
+
+	if err := insertIfNotExists(b.db, seoName, b.locales); err != nil {
+		panic(err)
+	}
+	return seo
+}
+
+// RemoveSEO removes the specified SEO,
+// if the SEO has children, the parent of the children will
+// be the parent of the SEO
+func (b *Builder) RemoveSEO(obj interface{}) *Builder {
+	seoToBeRemoved := b.GetSEO(obj)
+	if seoToBeRemoved == nil || seoToBeRemoved == b.seoRoot {
+		return b
+	}
+	seoToBeRemoved.removeSelf()
+	delete(b.registeredSEO, seoToBeRemoved.name)
+	return b
+}
+
+// GetSEO retrieves the specified SEO, It accepts two types of parameters.
+// One is a string, where the literal value of the parameter is the name of the SEO.
+// The other is an instance of a struct embedded with the Setting type, in which case
+// the SEO name is obtained from the type name that is retrieved through reflection.
+// If no SEO with the specified name is found, it returns nil.
+func (b *Builder) GetSEO(obj interface{}) *SEO {
+	switch res := obj.(type) {
+	case string:
+		return b.registeredSEO[res]
+	default:
+		modelType := reflect.Indirect(reflect.ValueOf(obj)).Type()
+		return b.registeredSEO[modelType]
+	}
+}
+
+func (b *Builder) GetGlobalSEO() *SEO {
+	return b.seoRoot
+}
+
+// GetSEOPriority gets the priority of the specified SEO,
+// with higher number indicating higher priority.
+// The priority of Global SEO is 1 (the lowest priority)
+func (b *Builder) GetSEOPriority(name string) int {
+	node := b.GetSEO(name)
+	depth := 0
+	for node != nil && node.name != "" {
+		node = node.parent
+		depth++
+	}
+	return depth
+}
+
+func (b *Builder) SortSEOs(SEOs []*QorSEOSetting) {
+	orders := make(map[string]int)
+	order := 0
+	var dfs func(root *SEO)
+	dfs = func(seo *SEO) {
+		if seo == nil {
+			return
+		}
+		orders[seo.name] = order
+		order++
+		for _, child := range seo.children {
+			dfs(child)
+		}
+	}
+	dfs(b.seoRoot)
+	sort.Slice(SEOs, func(i, j int) bool {
+		return orders[SEOs[i].Name] < orders[SEOs[j].Name]
+	})
+}
+
+// AfterSave sets the hook called after saving
+func (b *Builder) AfterSave(v func(ctx context.Context, settingName string, locale string) error) *Builder {
+	b.afterSave = v
+	return b
+}
+
+func (b *Builder) Render(obj interface{}, req *http.Request) h.HTMLComponent {
+	seo := b.GetSEO(obj)
+	if seo == nil {
+		return h.RawHTML("")
+	}
+
+	locale := defaultLocale
+	if v, ok := obj.(l10n.L10nInterface); ok {
+		if v.GetLocale() != "" {
+			locale = v.GetLocale()
+		}
+	}
+	localeFinalSeoSetting := seo.getLocaleFinalQorSEOSetting(locale, b.db)
+	return b.render(obj, localeFinalSeoSetting, seo, req)
+}
+
+// BatchRender rendering multiple SEOs at once.
+// It is the responsibility of the caller to ensure that every element in objs
+// is of the same type, as it is performance-intensive to check whether each element
+// in `objs` if of the same type through reflection.
+func (b *Builder) BatchRender(objs []interface{}, req *http.Request) []h.HTMLComponent {
```

Comment:
please change the type of `objs` to `interface{}` to accept any type of slice

Reply:
It has been changed successfully.